Ionyshch/upload buffer

ionyshch-prismsl opened the pull request.

ionyshch-prismsl committed:
Fixed collections names and build files according to review comments

ionyshch-prismsl committed:
Removed QtCreator files

ionyshch-prismsl committed:
Merge fixed from YuriyS for flipbook times.
Fixed JsonValue

ionyshch-prismsl committed:
Added min required boost version. Added using BOOST_ROOT environment variable

ionyshch-prismsl committed:
Made Boost link statically

ionyshch-prismsl committed:
In progress implementing file upload from memory buffer

ionyshch-prismsl committed:
Implemented background file upload from memory buffer

ionyshch-prismsl committed:
Merge branch 'master' into ionyshch/upload-buffer

ionyshch-prismsl committed:
Made client methods, that accepted path to image file, accept Payload instance.
Payload instance either provides file name to upload or memory buffer, that will
be uploaded with "file upload"

andrii-prism commented:
Please keep example camera code as small as possible. You can dup all code into "test/testcamera.cpp" or similar if you need.

Yurmusp commented on include/curl-session.h:
```diff
@@ -78,7 +78,8 @@ class CurlSession : public CurlCallbacks
 
 
-    void addFormFile(const std::string& key, const std::string& filePath, const std::string& mimeType);
+    void addFormFile(const char* key, const char* filePath, const char* mimeType);
```
typically, when you have std::string version, you do not need const char* version. 
It shall deduce to std::string automatically.

ionyshch-prismsl replied:
The idea is to avoid creating temp std::string object just to take its buffer with c_str() inside the method. I am not ready to say, if compiler can do such optimization and avoid creating string object.

Yurmusp replied:
Even if it does not, std::strings are pretty efficient. I do not think this is critical optimization. It just C style interface now. Not C++. I do not really understand why you moved to it.

ionyshch-prismsl replied:
One should you std::string unless there is a good reason not to do it. My point is that allocating/copying/deallocating memory on an (embedded) camera is the thing to avoid and if I know I have many calls that just pass const char* I'm accounting for it.
Of course, to get the most from this approach, it should be made everywhere. I just illustrated the point in one place. I would consider this an optimization and would agree it may be too early for such kind optimizations.
And instead of discussion, if we should avoid allocate/copy/deallocate and if yes, how to do it, I read comment about C style vs C++ style.
I don't understand the purpose of this comment.

Yurmusp replied:
Purpose of comment to keep code consistent if it does not contradict performance requirements. If we do C++ style we shall keep C++ style where possible. C style looks like inconsistency. I do not see performance penalty here.

ionyshch-prismsl replied:
I've come with a solution that both keeps C++ style and avoids creating temporary std::string object in case we have C string.

Yurmusp commented on include/domain-types.h:
```diff
@@ -121,6 +121,28 @@ struct Tag
+    }
+
+    std::string fileName;
+    const void* data;
```
Who owns this data?
I would rather do std::vector<uint8_t> instead. It is extra copy, but jpeg compression, it is nothing.

ionyshch-prismsl replied:
We don't transfer ownership. Caller own the data. Copying also means memory allocation and deallocation. We want to avoid those.
In case we really need to do a copy, it can be done inside method implementation, no need to made the choice in the interface.

Yurmusp replied:
To be clean implementation, then your caller shall create
shared_array pointer (instead of std::vector) and your object shall contain shared_array pointer. Then you have no copying and no pointers with unknown ownership.

ionyshch-prismsl replied:
There are several points here:
1. Caller doesn't always have a vector. There may just buffer he/she doesn't own, provided as pointer plus size. My approach covers this case too, while allowing input data to reside in a vector.
2. When method/function signature is const void* buf, size_t bufSize the ownership transfer is described in comments, and usually it isn't transferred as it incurs memory/performance penalty. Will providing comment for data and dataSize fields close the discussion?
3. I totally disagree about using shared_array as described. I think shared_ptr/array is misused most of the time, as shared means two or more owners that may (but not necessarily will) work with shared data simultaneously. Otherwise one would use unique_ptr. With shared one adds more confusion as it hides actual ownership transfer or its absence. 
4. Why do you suggest smart pointer here vs const void*? What problem does it solve? For me smart pointer is a design pattern, which, by definition, is used to solve particular problem in a (proper) context.
5. I try to avoid complex data types in the interface (std::string being an exception). I don't understand a desire to pass data structures with allocators through (potential) module boundary (as an alternative consider using pure C interface as libcurl has) to then deal with problems such as those you described with Windows dlls. Note, that dealing with problems means ugly solutions that are arguably better, than C-style approach.

Yurmusp replied:
2. Yes, comment would be much better, if we leave C style pointer.
3. I also thought about unique_ptr initially. However, I'm not sure if it works with array_ptr. I afraid it is not that's why I suggested array_ptr.
4. Only big issue here is ownership. The smaller issue is style, but we can ignore it if that solves the bigger one.
5. We already decided to ignore module boundary problems (like DLL related).

To keep interface universal, we can keep C style pointer, but you shall have comment that it does not own that memory area.

ionyshch-prismsl replied:
OK

Yurmusp commented on examples/yt_camera/main.cpp:
```diff
@@ -275,7 +281,21 @@ int main(int argc, char** argv)
+                                                   os, prc::Payload(BLOB_TMP_FILE));
+#else
+                std::vector<uchar> buf;
+                bool rv = imencode(".jpg", blob, buf, compression_params);
```
Jpeg is via OpenCV now. Not sure it will be best option. We probably will have to go turbojpeg or think if camera provides some custom HW supported jpeg compression.

ionyshch-prismsl replied:
HW supported, if available would be the best.
For now, consider this example as example for PC, because for embedded we'd like to minimize number of third-party libraries built to ease SDK use.

Yurmusp commented on src/client.cpp:
```diff
@@ -393,7 +393,14 @@ status_t Client::Impl::uploadBackground(id_t accountId, id_t instrumentId,
-    cs->addFormFile(kStrData, imageFile, mimeTypeFromFilePath(imageFile));
+
+    if (payload.data)
+        cs->addFormFile(kStrData, payload.data, payload.dataSize, payload.mimeType.c_str());
```
We need to make sure we avoid Base64 encoding which brings 30% size penalty.
If we do not, we shall think how to do it without it.
See also some discussion here.http://blog.marcinbudny.com/2014/02/sending-binary-data-along-with-rest-api.html#.WHgXwKIrKIZ
If we have base64 in current impl, we probably shall go first way in this article.
We need to change Connect API on web backend, but since we are asked to be super efficient, we have to do it earlier or later. So, we shall either do it or prepare well for it.
Lets discuss.

Yurmusp replied:
I also talked to Keith and he says that for frames with many objects json data for object stream becomes an issue on lower bandwidth connections.
I think for now we will stay with json, but for the future, we will need to use protobuf. We need web API to be ready for that though. So, now we do json.

ionyshch-prismsl replied:
Image is sent as binary, not Base64. We may try to request compressing JSON and see, whether it solves an issue.
Will switch to protobuf (what are alternatives?) later

Yurmusp replied:
If it is sent as binary that does not go via Base64 that's good. 
How you know it is sent like it?

ionyshch-prismsl replied:
curl with --trace parameter; logging size of image and uploaded size.

Yurmusp replied:
I do not fully understand if it really confirms the statement that it goes binary, but if you absolutely sure then let it be.

ionyshch-prismsl replied:
"The encoding type application/x-www-form-urlencoded is inefficient for sending large quantities of binary data or text containing non-ASCII characters.  Thus, a new media type, multipart/form-data, is proposed as a way of efficiently sending the values associated with a filled-out form from client to server." [see](https://tools.ietf.org/html/rfc1867)

Yurmusp commented on examples/yt_camera/main.cpp:
```diff
@@ -349,18 +369,33 @@ int main(int argc, char** argv)
+                                                 prc::Payload(BACKGROUND_TMP_FILE));
+#else
+                std::vector<uchar> buf;
+                bool rv = imencode(".jpg", background, buf, compression_params);
```
This is OpenCV jpeg encoding, right?
What it uses inside for it? Libjpeg?
We have to be max efficient.
We can have this in initial impl for demos, but eventually, we shall move to something super optimized.
Turbojpeg lib or HW encoder.

ionyshch-prismsl committed:
Created separate app test-client for testing. It differs from examples/yt_camera in extra code
to control how client methods are called, extra output.

Cleaned ty_camera from testing code to make it more compact.

Made client accept video files as Payload, allowing upload from memory buffer.

ionyshch-prismsl committed:
Added wrapper class that allows passing both const char* and string into (internal)
methods without creating temporary object (as it would be if function accepts std::string
and caller has C-string e.g. const like "const string").

Added comment about data ownership and availability (life time) for Payload's data parameter.
